metadata-builders/src/main/java/com/linkedin/metadata/builders/search/DatasetIndexBuilder.java

```diff
@@ -1,35 +1,84 @@
+import com.linkedin.restli.client.Client;
 import com.linkedin.schema.SchemaField;
 import com.linkedin.schema.SchemaMetadata;
 import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
+import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
 import javax.annotation.Nonnull;
+import lombok.SneakyThrows;
 import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+
 
 @Slf4j
 public class DatasetIndexBuilder extends BaseIndexBuilder<DatasetDocument> {
-  public DatasetIndexBuilder() {
+  private static RestliRemoteDAO<DataPlatformSnapshot, DataPlatformAspect, DataPlatformUrn> dataPlatformDAO;
+
+  /**
+   * Constructor for invoking dataset search index builder.
+   *
+   * @param restliClient restli client used to interact with other services to build dataset search documents.
+   */
+  public DatasetIndexBuilder(@Nonnull Client restliClient) {
     super(Collections.singletonList(DatasetSnapshot.class), DatasetDocument.class);
+    dataPlatformDAO = new RestliRemoteDAO<>(DataPlatformSnapshot.class, DataPlatformAspect.class, restliClient);
   }
 
+  private static final LoadingCache<DataPlatformUrn, String> DELIMITER_CACHE =
+      CacheBuilder.newBuilder().maximumSize(1000) // maximum 1000 records in cached
+          .expireAfterWrite(1, TimeUnit.DAYS) // cache expire after one day since write
+          .build(new CacheLoader<DataPlatformUrn, String>() {
+            @Override
+            public String load(@Nonnull DataPlatformUrn platformUrn) {
+              log.info("checking gms to get delimiter for platform {}", platformUrn);
+
+              return dataPlatformDAO.get(DataPlatformInfo.class, platformUrn)
+                  .map(DataPlatformInfo::getDatasetNameDelimiter)
+                  .orElse(StringUtils.EMPTY);
+            }
+          });
+
+  @SneakyThrows
   @Nonnull
-  private static String buildBrowsePath(@Nonnull DatasetUrn urn) {
-    return ("/" + urn.getOriginEntity() + "/"  + urn.getPlatformEntity().getPlatformNameEntity() + "/" + urn.getDatasetNameEntity())
-        .replace('.', '/').toLowerCase();
+  static String buildBrowsePath(@Nonnull DatasetUrn urn) {
+    final String dataOrigin = urn.getOriginEntity().name();
+    final String platform = urn.getPlatformEntity().getPlatformNameEntity();
+    final String dataset = urn.getDatasetNameEntity();
+    String browsePath = "/" + dataOrigin + "/" + platform + "/" + dataset;
+    final String delimiter = DELIMITER_CACHE.get(urn.getPlatformEntity());
+    log.debug("delimiter for platform {}, is {} ", urn.getPlatformEntity().getPlatformNameEntity(), delimiter);
+    if (!delimiter.isEmpty()) {
+      final Character delimiterChar = delimiter.charAt(0);
+      if (delimiterChar.equals('/')) {
+        browsePath = "/" + dataOrigin + "/" + platform + dataset;
+      } else {
+        browsePath = ("/" + dataOrigin + "/" + platform + "/" + dataset).replace(delimiterChar, '/');
```

Comment:
Shouldn't this be `... + dataset.replace(delimiterChar, '/')`, not the entire thing?

Reply:
So might be worth adding a test here too.

If my platform is something like `aws:s3`, delimiter `:`, and dataset `foo:bar`, what would this return? I'd assume `/origin/aws:s3/foo/bar` is expected, but this returns `/origin/aws/s3/foo/bar`.